pymc3/distributions/multivariate.py

```diff
@@ -737,6 +738,21 @@ def __str__(self):
 matrix_pos_def = PosDefMatrix()
 
 
+class WishartRV(RandomVariable):
+    name = "wishart"
+    ndim_supp = 0
```

Comment:
Shouldn't this be `ndim_supp = 2` (i.e. the Wishart distribution is over matrices)?

Reply:
It somehow fails if we set `ndim_supp = 2`. Strangely, the samples returned by `rng_fn` are actually 2-dimensional. Any ideas why this might be happening.

Reply:
You might need to override `_shape_from_params` as is done here, if the RV is failing to guess the shape with the default method: https://github.com/aesara-devs/aesara/blob/3a3adaee3b8ceccff6d0208996f117fcd5e5eb6f/aesara/tensor/random/basic.py#L518-L521

What error are you seeing?

Reply:
> What error are you seeing?

<details>
<summary>Here's the full traceback</summary>
<br>





```
________________________________________________________ TestWishart.test_distribution ________________________________________________________

self = <pymc3.tests.test_distributions_random.TestWishart object at 0x7f53e0b50810>

    def test_distribution(self):
        self.validate_tests_list()
>       self._instantiate_pymc_rv()

pymc3/tests/test_distributions_random.py:300: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pymc3/tests/test_distributions_random.py:311: in _instantiate_pymc_rv
    **params, size=self.size, rng=aesara.shared(self.get_random_state(reset=True))
pymc3/distributions/multivariate.py:816: in dist
    return super().dist([nu, V], *args, **kwargs)
pymc3/distributions/distribution.py:285: in dist
    rv_out = cls.rv_op(*dist_params, size=create_size, **kwargs)
../../../miniconda3/envs/pymc3-dev-py37/lib/python3.7/site-packages/aesara/tensor/random/op.py:323: in __call__
    res = super().__call__(rng, size, dtype, *args, **kwargs)
../../../miniconda3/envs/pymc3-dev-py37/lib/python3.7/site-packages/aesara/graph/op.py:269: in __call__
    node = self.make_node(*inputs, **kwargs)
../../../miniconda3/envs/pymc3-dev-py37/lib/python3.7/site-packages/aesara/tensor/random/op.py:370: in make_node
    bcast = self.compute_bcast(dist_params, size)
../../../miniconda3/envs/pymc3-dev-py37/lib/python3.7/site-packages/aesara/configparser.py:49: in res
    return f(*args, **kwargs)
../../../miniconda3/envs/pymc3-dev-py37/lib/python3.7/site-packages/aesara/tensor/random/op.py:287: in compute_bcast
    shape = self._infer_shape(size, dist_params)
../../../miniconda3/envs/pymc3-dev-py37/lib/python3.7/site-packages/aesara/tensor/random/op.py:257: in _infer_shape
    param_shapes=param_shapes,
pymc3/distributions/multivariate.py:750: in _shape_from_params
    self.ndim_supp, dist_params, rep_param_idx, param_shapes
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

ndim_supp = 2, dist_params = (TensorConstant{4}, TensorConstant{[[1. 0. 0...0. 0. 1.]]}), rep_param_idx = 2
param_shapes = [(), (Subtensor{int64}.0, Subtensor{int64}.0)]

    def default_shape_from_params(
        ndim_supp, dist_params, rep_param_idx=0, param_shapes=None
    ):
        """Infer the dimensions for the output of a `RandomVariable`.
    
        This is a function that derives a random variable's support
        shape/dimensions from one of its parameters.
    
        XXX: It's not always possible to determine a random variable's support
        shape from its parameters, so this function has fundamentally limited
        applicability and must be replaced by custom logic in such cases.
    
        XXX: This function is not expected to handle `ndim_supp = 0` (i.e.
        scalars), since that is already definitively handled in the `Op` that
        calls this.
    
        TODO: Consider using `aesara.compile.ops.shape_i` alongside `ShapeFeature`.
    
        Parameters
        ----------
        ndim_supp: int
            Total number of dimensions for a single draw of the random variable
            (e.g. a multivariate normal draw is 1D, so `ndim_supp = 1`).
        dist_params: list of `aesara.graph.basic.Variable`
            The distribution parameters.
        param_shapes: list of tuple of `ScalarVariable` (optional)
            Symbolic shapes for each distribution parameter.  These will
            be used in place of distribution parameter-generated shapes.
        rep_param_idx: int (optional)
            The index of the distribution parameter to use as a reference
            In other words, a parameter in `dist_param` with a shape corresponding
            to the support's shape.
            The default is the first parameter (i.e. the value 0).
    
        Results
        -------
        out: a tuple representing the support shape for a distribution with the
        given `dist_params`.
    
        """
        if ndim_supp <= 0:
            raise ValueError("ndim_supp must be greater than 0")
        if param_shapes is not None:
>           ref_param = param_shapes[rep_param_idx]
E           IndexError: list index out of range

../../../miniconda3/envs/pymc3-dev-py37/lib/python3.7/site-packages/aesara/tensor/random/op.py:69: IndexError
```
</details>

And overriding `_shape_from_params` didn't work. You might also wanna take a look at the test first, just to see if the parameters are valid.

Reply:
It's definitely looks like you need to override that method. I didn't imply that it should be the same overwrite, just that it needs one.

Reply:
> It somehow fails if we set `ndim_supp = 2`. Strangely, the samples returned by `rng_fn` are actually 2-dimensional. Any ideas why this might be happening?

`ndim_supp` is used for shape inference and a few other related things, but not necessarily for critical `RandomVariable.perform` operations, so it's not surprising that this value would be incorrect and still produce valid samples.

Regardless, we need the `ndim_supp` values to be accurate.  We might also need to add some more sanity checks at the symbolic level using that property (i.e. when/if we can verify that the output dimensions match `ndim_supp`, we probably should).

Reply:
Otherwise, @ricardoV94 is correct; it looks like this needs a custom `_shape_from_params` that says the second parameter, `V`, determines the output shape.  I think the default assumes that the first parameter does, which is clearly wrong in this case.

Reply:
The current approach should cover the correct `ndim_supp` value as well as the returned shape correctly. I simply returned the shape of `V` from the `param_shapes` list.